codec/src/main/java/io/netty/handler/codec/compression/ZstdEncoder.java

```diff
@@ -0,0 +1,193 @@
+        }
+
+        if (buffer == null) {
+            buffer = ctx.alloc().directBuffer(blockSize);
+        }
+
+        long targetBufSize = 0;
+        int remaining = msg.readableBytes() + buffer.readableBytes();
+
+        // quick overflow check
+        if (remaining < 0) {
+            throw new EncoderException("too much data to allocate a buffer for compression");
+        }
+
+        while (remaining > 0) {
+            int curSize = Math.min(blockSize, remaining);
+            remaining -= curSize;
+            targetBufSize += Zstd.compressBound(curSize);
+        }
+
+        if (targetBufSize > maxEncodeSize || 0 > targetBufSize) {
+            throw new EncoderException("requested encode buffer size (" + targetBufSize + " bytes) exceeds " +
+                    "the maximum allowable size (" + maxEncodeSize + " bytes)");
+        }
+
+        int bufferSize = (int) targetBufSize;
+        return ctx.alloc().directBuffer(bufferSize, bufferSize);
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) {
+
+        if (buffer == null) {
+            throw new IllegalStateException("not added to a pipeline," +
+                    "or has been removed,buffer is null");
+        }
+
+        final ByteBuf buffer = this.buffer;
+        int length;
+        while ((length = in.readableBytes()) > 0) {
+            final int nextChunkSize = Math.min(length, buffer.writableBytes());
+            in.readBytes(buffer, nextChunkSize);
+
+            if (!buffer.isWritable()) {
+                flushBufferedData(out);
+            }
+        }
+    }
+
+    private void flushBufferedData(ByteBuf out) {
+
+        final int flushableBytes = buffer.readableBytes();
+        if (flushableBytes == 0) {
+            return;
+        }
+
+        final int bufSize = (int) Zstd.compressBound(flushableBytes);
+        out.ensureWritable(bufSize);
+        final int idx = out.writerIndex();
+        int compressedLength;
+        try {
+            ByteBuffer outNioBuffer = out.internalNioBuffer(idx, out.writableBytes());
+            compressedLength = Zstd.compress(
+                    outNioBuffer,
+                    buffer.internalNioBuffer(buffer.readerIndex(), flushableBytes),
+                    compressionLevel);
+        } catch (Exception e) {
+            throw new CompressionException(e);
+        }
+
+        out.writerIndex(idx + compressedLength);
+        buffer.clear();
+    }
+
+    @Override
+    public void flush(final ChannelHandlerContext ctx) {
+        if (buffer != null && buffer.isReadable()) {
+            final ByteBuf buf = allocateByteBuffer(ctx, Unpooled.EMPTY_BUFFER, isPreferDirect());
+            flushBufferedData(buf);
+            ctx.write(buf);
+        }
+        ctx.flush();
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) {
+        buffer = ctx.alloc().directBuffer(blockSize);
+        buffer.clear();
```

Comment:
this call is not needed.

Reply:
The allocateByteBuffer method no longer allocates memory to the buffer, so it still allocates memory when the handlerAdded is called.